### PR TITLE
api: Always make subsampling factor symbolic

### PR DESCRIPTION
**Pull request by mloubout**
Fix issue with overwrties with different factor

### CODE REVIEW

File: devito/types/dense.py [hunk 1]
```diff
@@ -793,8 +793,8 @@ def _arg_defaults(self, alias=None):
         args = ReducerMap({key.name: self._data_buffer})
 
         # Collect default dimension arguments from all indices
-        for i, s in zip(self.dimensions, self.shape):
-            args.update(i._arg_defaults(_min=0, size=s))
+        for a, i, s in zip(key.dimensions, self.dimensions, self.shape):
```
Comment (mloubout) [inline devito/types/dense.py:796]: @FabioLuporini quite surprised this never broke anything

Comment (FabioLuporini) [reply devito/types/dense.py:796]: I'm now confused. 

Shouldn't it be `key._data_buffer`?
Shouldn't it be `key.shape`?

Might just be early morning sleepy brain

File: devito/types/dimension.py [hunk 2]
```diff
@@ -890,7 +900,7 @@ def _arg_defaults(self, _min=None, size=None, alias=None):
             return defaults
         try:
             # Is it a symbolic factor?
-            factor = defaults[dim._factor.name] = dim._factor.data
+            factor = defaults[dim._factor.name] = self._factor.data
```
Comment (FabioLuporini) [inline devito/types/dimension.py:903]: why?

File: devito/types/dimension.py [hunk 3]
```diff
@@ -848,13 +849,22 @@ def __init_finalize__(self, name, parent=None, factor=None, condition=None,
 
         super().__init_finalize__(name, parent)
 
-        self._factor = factor
+        # Always make the factor symbolic to allow overrides with different factor.
```
Comment (FabioLuporini) [inline devito/types/dimension.py:852]: We totally sure we don't want to leave this to the user? 

Probably yes for the principle of least surprise right? the underlying issue is when you pass an override with a TimeFunction with a different factor -- correct?

A symbolic factor will make the generated code a bit more verbose (thinking about when it's used in tandem with lazy streaming), but yeah, it sounds safer...

File: devito/types/dense.py (hunk 1 above)
Comment (mloubout) [reply devito/types/dense.py:796]: No:
- `key/alias` is the object and names used to build the operator, i.e the "key" of key-value pairs
- `self` is the runtime argument with the runtime shapes/values/data/....

So you want to return the {key: self} key-pair values

File: devito/types/dimension.py (hunk 2 above)
Comment (mloubout) [reply devito/types/dimension.py:903]: same as above, self is the runtime argument with the value for the operator, dim is the name of the dimension used to build the operator and argument name

File: devito/types/dimension.py (hunk 3 above)
Comment (mloubout) [reply devito/types/dimension.py:852]: I think it's safer to make it symbolic. It makes it a little it more verbose yes but at the price of avoiding bad surprises.

File: devito/types/dimension.py [hunk 4]
```diff
@@ -848,13 +849,22 @@ def __init_finalize__(self, name, parent=None, factor=None, condition=None,
...
+        if factor is None:
+            self._factor = None
+        elif is_integer(factor):
+            self._factor = Constant(name="%sf" % name, value=factor, dtype=np.int32)
+        elif factor.is_Constant and is_integer(factor.data):
```
Comment (FabioLuporini) [inline devito/types/dimension.py:857]: issubclass(factor, np.integer)

slightly cleaner

File: devito/types/dense.py (hunk 1 above)
Comment (FabioLuporini) [reply devito/types/dense.py:796]: ah yes you're right of course

**Pull request merged**